src/MongoDB.Driver/Linq/Translators/PredicateTranslator.cs

```diff
@@ -45,6 +43,8 @@ public static BsonDocument Translate<TDocument>(Expression<Func<TDocument, bool>
             var context = new PipelineBindingContext(serializerRegistry);
             context.AddExpressionMapping(predicate.Parameters[0], parameterExpression);
 
+            ExpressionValidator.Validate(predicate);
+
```

Comment:
Don't we want to partially evaluate BEFORE validating?

Reply:
For now, I don't really understand what is the idea of the `PartialEvaluator` class.
But, we need to check the expression tree before any changes in the expression tree, since the validation rules are built based on the C# trees.
So there are two options here:
1. if `PartialEvaluator` changes an expression tree => we have to use `ExpressionValidator` befor4e `PartialEvaluator`.
2. if not so, I guess, it may be better to check the expression tree syntax before any other activities. But the calling order of this methods it doesn't really matter in this case.

Reply:
After a small discussion with @craig - I understood the idea of `PartialEvaluator`.
So, I will call `PartialEvaluator` firstly.

Reply:
Done